File: data structures and algorithms/cpp/06_020.cpp
#include <iostream>
#include <string>
#include <filesystem>

namespace fs = std::filesystem;

void explore_directory(const fs::path& path, int depth = 0) {
    if (!fs::is_directory(path)) {
        return;
    }
    
    // Process current directory
    for (int i = 0; i < depth; i++) {
        std::cout << "  ";
    }
    std::cout << path.filename().string() << std::endl;
    
    // Explore subdirectories (DFS!)
    for (const auto& entry : fs::directory_iterator(path)) {
        explore_directory(entry.path(), depth + 1);
    }
}

File: data structures and algorithms/cs/06_020.cs
using System;
using System.IO;
using System.Linq;

public class DirectoryExplorer
{
   public void ExploreDirectory(string path, int depth)
   {
      // Validate directory path
      if (!Directory.Exists(path))
      {
         return;
      }

      // Display current directory with indentation
      Console.WriteLine(new string(' ', depth * 2) + new DirectoryInfo(path).Name);

      // Explore subdirectories using DFS approach
      try
      {
         foreach (var item in Directory.GetFileSystemEntries(path))
         {
            ExploreDirectory(item, depth + 1);
         }
      }
      catch (Exception e)
      {
         Console.Error.WriteLine("Error accessing directory: " + e.Message);
      }
   }
}
